ADR: Running out of Rancher Manager cluster (#95)
rancher-sandbox/rancher-turtles: docs/adr/0004-running-out-of-rancher-manager-cluster.md

# 4. Running out of Rancher Manager cluster

* Status: proposed
* Date: 2023-08-31
* Authors: @salasberryfin
* Deciders: @richardcase @alexander-demicev @furkatgofurov7 @Danil-Grigorev @mjura

## Context

> As an operator, I want the choice to deploy Rancher Turtles in the same or a different cluster to Rancher Manager, so that I have choice in my deployment topology.

Reviewer: ```suggestion
As an operator, I want the choice to deploy Rancher Turtles in the same or a different cluster to Rancher Manager, so that I have a choice in my deployment topology.
```

## Decision

`CAPIImportReconciler` will contain two differentiated clients:
- `Client`: in-cluster client for the controller and the CAPI resources.
- `RancherClient`: for the Rancher Manager cluster.

A flag `rancher-kubeconfig` with a path to a kubeconfig file is used to select the type of installation:

> - If no value is passed, Rancher Turtles is set to the same cluster installation: `Client` and `RancherClient` are the same instance of client.

Reviewer: ```suggestion
- If no value is passed, Rancher Turtles is set to the same cluster installation: `Client` and `RancherClient` are the same instance of the client.
```

- If a valid path to a kubeconfig file is passed, `RancherClient` is set to client created from the kubeconfig.

From an end-user perspective:

* No parameter is required if Rancher Manager and Rancher Turtles run in the same cluster.
* User will provide a path to a valid kubeconfig file that is available in the pod if installing Rancher Turtles outside of Rancher Manager cluster.

## Consequences

> - The path to the kubeconfig must be mounted in the pod to be accessible.

Reviewer 2: Might be worth adding that this is the pattern used by controller-runtime with the `--kubeconfig`

# 4. Running out of Rancher Manager cluster

* Status: proposed
* Date: 2023-08-31
* Authors: @salasberryfin
* Deciders: @richardcase @alexander-demicev @furkatgofurov7 @Danil-Grigorev @mjura

## Context

As an operator, I want the choice to deploy Rancher Turtles in the same or a different cluster to Rancher Manager, so that I have choice in my deployment topology.

> The initial implementation of Rancher Turtles assumes that the operator runs in the same cluster as Rancher Manager.

Reviewer 3: I think this could be removed. We are now making the initial version of turtles, and there is no previous ADR on this topic to replace.

## Decision

`CAPIImportReconciler` will contain two differentiated clients:

> - `Client`: for the controller itself for the CAPI Clusters.

Reviewer 3: ```suggestion
- `Client`: in-cluster client for the controller and the CAPI resources.
```

> - `RancherClient`: for Rancher.

Reviewer 3: ```suggestion
- `RancherClient`: for the Rancher Manager cluster.
```

A flag `rancher-kubeconfig` with a path to a kubeconfig file is used to select the type of installation:

> - If no value is passed, Rancher Turtles is set to a same cluster installation: `Client` and `RancherClient` are set to the same cluster.

Reviewer 3: ```suggestion
- If no value is passed, Rancher Turtles is set to the same cluster installation: `Client` and `RancherClient` are the same instance of client.
```

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [4. Running out of Rancher Manager cluster](#4-running-out-of-rancher-manager-cluster)
  - [Context](#context)
  - [Decision](#decision)
  - [Consequences](#consequences)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# 4. Running out of Rancher Manager cluster

* Status: proposed
* Date: 2023-08-31
* Authors: @salasberryfin
* Deciders: @richardcase @alexander-demicev @furkatgofurov7 @Danil-Grigorev @mjura

## Context

As an operator, I want the choice to deploy Rancher Turtles in the same or a 
different cluster to Rancher Manager, so that I have a choice in my deployment 

> opology.

Reviewer 4: ```suggestion
topology.
```
